File: python/tempmail_wrapper/providers/tempmail_lol.py
"""tempmail.lol provider implementation (REST API, token-based)."""

from __future__ import annotations

from datetime import datetime

from ..base import TempMailProvider
from ..exceptions import NotFoundError, TempMailError
from ..http_client import HttpClient
from ..models import Message, MessageDetail


class TempmailLolProvider(TempMailProvider):
    """Provider for tempmail.lol — REST API, token-based, no auth."""

    BASE_URL = "https://api.tempmail.lol/v2"

    def __init__(
        self,
        proxies: list[str] | None = None,
        random_ua: bool = True,
        **kwargs,
    ) -> None:
        self._http = HttpClient(proxies=proxies, random_ua=random_ua, use_cookies=False)
        self._token: str | None = None
        self._email: str | None = None

    @property
    def name(self) -> str:
        return "tempmail.lol"

    def generate_email(self) -> str:
        resp = self._http.post(
            f"{self.BASE_URL}/inbox/create",
            timeout=30,
            headers={"Content-Type": "application/json"},
        )
        if resp.status_code not in (200, 201):
            raise TempMailError(
                f"Failed to create email: {resp.status_code}", self.name
            )
        data = resp.json()
        self._email = data.get("address")
        self._token = data.get("token")
        if not self._email or not self._token:
            raise TempMailError("Missing address or token in response", self.name)
        return self._email

    def get_inbox(self, email: str) -> list[Message]:
        if not self._token:
            raise TempMailError("No token — call generate_email() first", self.name)
        resp = self._http.get(
            f"{self.BASE_URL}/inbox",
            params={"token": self._token},
            timeout=30,
        )
        if resp.status_code != 200:
            raise TempMailError(f"Inbox error: {resp.status_code}", self.name)
        data = resp.json()
        if data.get("expired"):
            raise TempMailError("Token expired", self.name)
        messages = []
        for item in data.get("emails", []):
            date_str = item.get("date", item.get("createdAt", ""))
            try:
                date = datetime.fromisoformat(date_str.replace("Z", "+00:00"))
            except (ValueError, AttributeError):
                date = datetime.now()
            messages.append(
                Message(
                    id=str(item.get("_id", item.get("id", item.get("uid", "")))),
                    sender=item.get("from", item.get("sender", "")),
                    subject=item.get("subject", ""),
                    date=date,
                )
            )
        return messages

    def read_message(self, message_id: str) -> MessageDetail:
        if not self._token:
            raise TempMailError("No token — call generate_email() first", self.name)
        # tempmail.lol returns full emails in the inbox response
        resp = self._http.get(
            f"{self.BASE_URL}/inbox",
            params={"token": self._token},
            timeout=30,
        )
        if resp.status_code != 200:
            raise NotFoundError(f"Cannot read inbox: {resp.status_code}", self.name)
        data = resp.json()
        for item in data.get("emails", []):
            if str(item.get("_id", item.get("id", item.get("uid", "")))) == str(message_id):
                date_str = item.get("date", item.get("createdAt", ""))
                try:
                    date = datetime.fromisoformat(date_str.replace("Z", "+00:00"))
                except (ValueError, AttributeError):
                    date = datetime.now()
                return MessageDetail(
                    id=str(message_id),
                    sender=item.get("from", item.get("sender", "")),
                    subject=item.get("subject", ""),
                    date=date,
                    body_text=item.get("body", item.get("text", "")),
                    body_html=item.get("html", ""),
                    attachments=[],
                )
        raise NotFoundError(f"Message {message_id} not found", self.name)

    def delete_email(self, email: str) -> bool:
        self._token = None
        self._email = None
        return True

File: rust/src/providers/tempmail_lol.rs
use async_trait::async_trait;
use serde::Deserialize;

use crate::error::TempMailError;
use crate::httpclient::HttpClient;
use crate::models::{Message, MessageDetail};
use crate::providers::TempMailProvider;

const BASE_URL: &str = "https://api.tempmail.lol/v2";

/// Provider for tempmail.lol — REST API, token-based, no auth header.
#[derive(Debug)]
pub struct TempmailLol {
    client: HttpClient,
    /// (email, token) cached after generation.
    state: std::sync::Arc<tokio::sync::Mutex<Option<(String, String)>>>,
}

#[derive(Deserialize)]
struct CreateResp {
    address: Option<String>,
    token: Option<String>,
}

#[derive(Deserialize)]
struct InboxResp {
    #[serde(default)]
    expired: bool,
    #[serde(default)]
    emails: Vec<serde_json::Value>,
}

impl TempmailLol {
    pub fn new(client: HttpClient) -> Self {
        Self {
            client,
            state: std::sync::Arc::new(tokio::sync::Mutex::new(None)),
        }
    }

    async fn token(&self) -> Result<String, TempMailError> {
        self.state
            .lock()
            .await
            .clone()
            .map(|(_, t)| t)
            .ok_or_else(|| TempMailError::Api("Must call generate_email first".into()))
    }

    async fn fetch_inbox(&self) -> Result<Vec<serde_json::Value>, TempMailError> {
        let token = self.token().await?;
        let resp: InboxResp = self
            .client
            .get_json(&format!("{}/inbox?token={}", BASE_URL, token))
            .await?;
        if resp.expired {
            return Err(TempMailError::Api("Token expired".into()));
        }
        Ok(resp.emails)
    }
}

fn parse_iso(s: &str) -> chrono::DateTime<chrono::Utc> {
    chrono::DateTime::parse_from_rfc3339(s)
        .map(|dt| dt.with_timezone(&chrono::Utc))
        .unwrap_or_else(|_| chrono::Utc::now())
}

fn opt_str(v: Option<&serde_json::Value>) -> String {
    v.and_then(|x| x.as_str()).unwrap_or("").to_string()
}

fn opt_vid(v: Option<&serde_json::Value>) -> String {
    match v {
        Some(serde_json::Value::String(s)) => s.clone(),
        Some(serde_json::Value::Number(n)) => n.to_string(),
        Some(other) => other.to_string(),
        None => String::new(),
    }
}

fn email_to_message(m: &serde_json::Value) -> Message {
    Message {
        id: opt_vid(m.get("_id").or_else(|| m.get("id")).or_else(|| m.get("uid"))),
        sender: opt_str(m.get("from").or_else(|| m.get("sender"))),
        subject: opt_str(m.get("subject")),
        date: parse_iso(&opt_str(m.get("date").or_else(|| m.get("createdAt")))), preview: String::new(), has_attachments: false }
}

#[async_trait]
impl TempMailProvider for TempmailLol {
    async fn generate_email(&self) -> Result<String, TempMailError> {
        let body = serde_json::json!({});
        let resp: CreateResp = self
            .client
            .post_json(&format!("{}/inbox/create", BASE_URL), &body)
            .await?;
        let email = resp
            .address
            .ok_or_else(|| TempMailError::Api("Missing address in response".into()))?;
        let token = resp
            .token
            .ok_or_else(|| TempMailError::Api("Missing token in response".into()))?;
        let mut state = self.state.lock().await;
        *state = Some((email.clone(), token));
        Ok(email)
    }

    async fn get_inbox(&self, _email: &str) -> Result<Vec<Message>, TempMailError> {
        let emails = self.fetch_inbox().await?;
        Ok(emails.iter().map(email_to_message).collect())
    }

    async fn read_message(&self, message_id: &str) -> Result<MessageDetail, TempMailError> {
        // ponytail: tempmail.lol has no single-message endpoint; re-fetch inbox and find by id
        let emails = self.fetch_inbox().await?;
        for m in &emails {
            if opt_vid(m.get("_id").or_else(|| m.get("id")).or_else(|| m.get("uid"))) == message_id {
                return Ok(MessageDetail {
                    message: email_to_message(m),
                    body_text: {
                        let b = opt_str(m.get("body").or_else(|| m.get("text")));
                        if b.is_empty() { None } else { Some(b) }
                    },
                    body_html: {
                        let h = opt_str(m.get("html"));
                        if h.is_empty() { None } else { Some(h) }
                    },
                    attachments: vec![], ..Default::default()                 });
            }
        }
        Err(TempMailError::NotFound)
    }

    async fn delete_email(&self, _email: &str) -> Result<bool, TempMailError> {
        let mut state = self.state.lock().await;
        *state = None;
        Ok(true)
    }
}
